README.md

```diff
@@ -1,3 +1,94 @@
 # Language Features a la carte
 
 A tool checking that some specific language features are not used
+
+## How to use the tool
+Defining which constructs are allowed to be used in a program can be done in two ways:
+* Blacklist-based checking: specify the constructs that <b>should not</b> be used
+* Whitelist-based checking: specify the constructs that <b>are allowed</b> to be used
+
+### Blacklisting language features
+To check a program using a blacklist-based approach, instantiate a `BlacklistChecker` with
+`BlacklistRule`s that match the constructs that should not be allowed.
+`BlacklistRule`s can be either predefined ones, available in `BlacklistRules` (e.g.
+`BlacklistRules.NoVar`) or custom ones.
+
+E.g. to check that a program does not use `while` or `var`:
+```Scala
+import syntactic_checker.blacklist.{ BlacklistChecker, BlacklistRules }
+import scala.meta.dialects.Sbt1
+
+// example source code
+val sourceCodeString = """
+  |private val y: String = f(x)
+  |""".stripMargin
+
+val checker = BlacklistChecker(
+  BlacklistRules.NoVar,
+  BlacklistRules.NoWhile
+)
+
+checker.checkCodeString(dialect = Sbt1, sourceCodeString) // returns Valid
+```
+
+To define a custom `BlacklistRule`, create an
+object named following the name of your rule and let it extend `BlacklistRule` with as `checkFunc`
+a `PartialFunction` that returns a `BlackListViolation` when given a construct that should be blacklisted (for
+other constructs it should not be defined) and as `msg` the message that should
+be displayed to the user when a construct is found in the program that is forbidden by the rule.
+
+E.g. defining a rule that forbids the usage of `null` can be done as follows:
+```Scala
+case object NoNull extends BlacklistRule(checkFunc = {
+    case nullKw: Lit.Null => reportNull(nullKw)
+}, msg = "usage of null is forbidden")
+private def reportNull(kw: Tree) = BlacklistViolation(kw, NoNull)
+```
+(the `NoNull` rule of this example is actually implemented in `BlackListRules`, so in practice it should not
+be redefined)
+
+*Note:* the method `reportNull` is needed because simply writing `case nullKw: Lit.Null => BlacklistViolation(nullKw, NoNull)`
+gives a compile error since `NoNull` cannot be referenced in its own definition.
```

Comment:
```Scala
case object NoNull extends BlacklistRule({
    case nullKw: Lit.Null => BlacklistViolation(nullKw, NoNull)
  }, msg = "usage of null is forbidden")
```
gives this error:
```
super constructor cannot be passed a self reference unless parameter is declared by-name
    case nullKw: Lit.Null => BlacklistViolation(nullKw, NoNull)
```
Maybe I should use a by-name parameter?

Reply:
It seems to be a limitation of Scala 2 :-/
It works fine in Scala 3: https://scastie.scala-lang.org/1gwaaJiuQnCy6X6nVbuStg
(the same example fails in Scala 2: https://scastie.scala-lang.org/vPGhdCq3RfGT0c9MtgjUlQ)

The only workaround I can think of would be to define `checkFunc` as an abstract member instead of a parameter:

~~~ scala
abstract class BlacklistFunc(val msg: String) {
  def checkFunc: PartialFunction[Tree, BlacklistViolation]
}

// And then
case object NoNull extends BlacklistRule("usage of null is forbidden") {
  val checkFunc = { case nullKw: Lit.Null => BlacklistViolation(nullKw, this) }
}
~~~

That being said, in `BlacklistViolation` we don’t really use the underlying `BlacklistRule`, we only use its `msg`, so, do we need to pass the rule instance at all? Or would the message be enough?

~~~ scala
case object NoNull extends BlacklistRule({
  case nullKw: Lit.Null => BlacklistViolation(nullKw, "usage of null is forbidden")
})
~~~

Reply:
Indeed, passing only the message is sufficient (at first I thought that I would have to use more info on the rule, but it is not the case). And this allows to merge both types of `Violation`s (because there is no need to store a reference to the violated rule anymore) and therefore removes the need for the type parameter in `Checker` and `CheckResult`.

Reply:
Changed in a16f9b7f09565673fb8e537ee81133fae347e0e3